kvclient: remove retries within sendToReplicas

andrewbaptist opened the pull request:
This change removes a retry loop within sendToReplicas that occured when
we hit a NLHE. We still retry within the outer loop of sendPartialBatch
and removing this also removes the only remaining reordering of the
replicas that are retrieved from the RangeCache.

There was a metric distsender.errors.inleasetransferbackoffs that tracks
the number of retries within this now removed loop that is also removed.

Epic: none

Release note: Removes the metric distsender.errors.inleasetransferbackoffs.

Commit:
kvclient: remove retries within sendToReplicas

This change removes a retry loop within sendToReplicas that occured when
we hit a NLHE. We still retry within the outer loop of sendPartialBatch
and removing this also removes the only remaining reordering of the
replicas that are retrieved from the RangeCache.

There was a metric distsender.errors.inleasetransferbackoffs that tracks
the number of retries within this now removed loop that is also removed.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117632)
<!-- Reviewable:end -->

Commit:
kvclient: remove deprecated metric

This commit removes a metric which counts the number of retries occur
due to NLHE. This loop is being removed in a future commit.

Epic: none

Release note (ops change): Removes the metric
distsender.errors.inleasetransferbackoffs.

Commit:
kvclient: remove unused MoveToFront

This commit removes ths MoveToFront method on the Transport. A major
advantage of this change is that the Transport no longer modifies the
underlying ReplicaSlice. This will allow a future commit to avoid
copying the slice to every transport.

Epic: none

Release note: None

andrewbaptist commented:
There are several commits here. The main one is "kvclient: remove retries within sendToReplicas". I split the other ones out to make this commit smaller. Let me know if you want me to walk through it with you.

arulajmani reviewed (commented):
This stuff is incredibly subtle and I might have missed some things. I think this PR would benefit from a second pair of eyes as well before we merge (cc @nvanbenschoten)

> Reviewed 3 of 3 files at r2, 2 of 2 files at r3, 3 of 3 files at r4, 4 of 4 files at r5, 6 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-:-NoU9yXi4x1Ic-kShRdw:b3sxq7d)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`-- commits` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoTzxIzACwdMeV7eIpf):*
Could you add more words about the motivation for this change, pitfalls of the previous approach, and why the newer approach is better? 

I'm having a hard time convincing myself that this change is copacetic. One concern I have is with the way the `EvictionToken` is passed from `sendPartialBatch` to `sendToReplicas`. This is done by value, so any changes to the leaseholder in `sendToReplicas` aren't going to be reflected in the token that'll be used to evict from the cache in `sendPartialBatch`. In cases where we receive a `sendError` or `RangeKeyMismatchError` from `sendToReplicas`, we want to evict from the cache regardless of whether the initial token was constructed with the freshest leaseholder or not. For all we know, `sendToReplicas` was informed of a new leaseholder, and returned a `sendError` or `RangeKeyMismatchError` despite that. What benefit is it to have `sendPartialBatch` construct a new transport and retry to fail again in such cases?
___
*[`-- commits` line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU3U5TChLw94hMc0U5):*
I don't think we should get rid of this metric. We should find room for it in the new structure. We've previously found this metric to be useful in support escalations where tail latency is high because a leaseholder is slow in applying its lease.
___
*[`-- commits` line 38 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU4OC-D3QnTEwA1IDV):*
See my previous comments -- we should still find a place for this metric.
___
*[`-- commits` line 51 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoUBc4a5OJwQ44K_6E8):*
We're using a sync.Pool for the replicas slice on the transport already. Do we still expect to see major benefits?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 251 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU4WhnE2515wNKst9o:-NoU4WhnE2515wNKst9p:b-5twvfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/67664ad7a73b36145e4c48c14794636ef225bb9a/pkg/kv/kvclient/kvcoord/dist_sender.go#L251)):*
> ```Go
> 	// The maximum number of times a replica is retried when it repeatedly returns
> 	// stale lease info.
> 	sameReplicaRetryLimit = 10
> ```

Getting rid of this seems like a downside of this approach. The rationale for this limit was explained in this comment: 

```
// If the leaseholder is the replica that we've just tried, and  
// we've tried this replica a bunch of times already, let's move on  
// and not try it again. This prevents us getting stuck on a replica  
// that we think has the lease but keeps returning redirects to us  
// (possibly because it hasn't applied its lease yet). Perhaps that  
// lease expires and someone else gets a new one, so by moving on we  
// get out of possibly infinite loops.
```

Are we no longer worried about this?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1892 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU5JBD6Ld2zfAzRCL3:-NoU5JBD6Ld2zfAzRCL4:b-wi834l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb0e6c36d20cc50636e36f3211a0f6ac6367e26/pkg/kv/kvclient/kvcoord/dist_sender.go#L1892)):*
> ```Go
> 					return response{pErr: kvpb.NewError(ambiguousErr)}
> 				} else {
> 					return response{pErr: pErr}
> ```

Isn't pErr nil here? Should this be err?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2550 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU73Qm0YYNi52MOFI-:-NoU73Qm0YYNi52MOFI0:br5jr0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb0e6c36d20cc50636e36f3211a0f6ac6367e26/pkg/kv/kvclient/kvcoord/dist_sender.go#L2550)):*
> "If not, we loop around and try the next replica"

Consider the following case: 

Range 10 has replicas on nodes 1, 2, 3, 4, 5; node 1 is the leaseholder but hasn't applied its lease yet. So, it continues to return the older leaseholder (say node 2) in its NLHE.

Say the client has updated leaseholder information. It contacts node 1 and gets stale lease information. It then ends up trying nodes 2, 3, 4, and 5 -- all of whom tell it that node 1 is the leaseholder. Finally, it exits this loop, evicts its range descriptor, and repeats the whole process again (until node 1 applies its lease). 

I think we're running into the issue https://github.com/cockroachdb/cockroach/pull/32877 was trying to solve again. Isn't the client better served simply backing off here, instead of trying nodes 2, 3, 4, and 5?
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1038 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoTj6JI85Vm5GuUN9vz:-NoTj6JI85Vm5GuUN9w-:b-2pfrtg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2c487e9e2533a0eea6b72a22f47e6367a4ebc3/pkg/kv/kvclient/rangecache/range_cache.go#L1038)):*
> ```Go
> }
> 
> // evictLocked evicts a cache entry unless its newer than the passed in
> ```

nit: s/its/it's/g
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1056 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU-snX0Lo7immOASES:-NoU-snX0Lo7immOASET:bm1rmx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2c487e9e2533a0eea6b72a22f47e6367a4ebc3/pkg/kv/kvclient/rangecache/range_cache.go#L1056)):*
> "the descriptors are the same"

Naviely, reading the code, one would assume the descriptor can be older as well. However that's not possible because the cache doesn't go backwards, and the supplied descriptor also came from the cache. So it must be equal to the one in the cache. This is explained in the comment below, but we should move that comment above this one.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1058 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoTtV6SEZHdm8Z5CDpx:-NoTtV6SEZHdm8Z5CDpy:bv468li) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2c487e9e2533a0eea6b72a22f47e6367a4ebc3/pkg/kv/kvclient/rangecache/range_cache.go#L1058)):*
> ```Go
> 	// The cached lease is newer and the descriptors are the same, so don't evict.
> 	if cachedEntry.lease.Sequence > etSequence {
> 		// if the cached entry has the same descriptor but the cached lease is newer, don't evict.
> ```

nit: if/If/g and wrap comments to 80 chars.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 259 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU9Im48-k33LwzHe8G:-NoU9Im48-k33LwzHe8H:b-tuzh0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/67664ad7a73b36145e4c48c14794636ef225bb9a/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L259)):*
> ```Go
> 	require.Falsef(t, errors.HasAssertionFailure(err),
> 		"expected no AssertionFailedError due to sanity check on transaction already committed")
> 	require.ErrorContainsf(t, aErr, reason,
> ```

Is this happening because we're swallowing an error and not correctly wrapping it somewhere? 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 841 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU3rY5CX44FbyAVqxf:-NoU3rY5CX44FbyAVqxg:bxj5o4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/67664ad7a73b36145e4c48c14794636ef225bb9a/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L841)):*
> ```Go
> 				t.Fatalf("%d: unexpected error: %v", i, pErr)
> 			}
> 			if got := ds.Metrics().InLeaseTransferBackoffs.Count(); got != c.expected {
> ```

Is this test meaningful if we're getting rid of this? 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 919 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU4BjYCH5egF-2Bll6:-NoU4BjYCH5egF-2Bll7:b-6hm9u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/67664ad7a73b36145e4c48c14794636ef225bb9a/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L919)):*
> ```Go
> 	require.Nil(t, pErr)
> 	require.Equal(t, []roachpb.NodeID{1, 2}, sentTo)
> 	require.Equal(t, int64(0), ds.Metrics().InLeaseTransferBackoffs.Count())
> ```

Once we find a place for this metric in the new structure, keeping this test around with the assertion will be useful in convincing us we're doing the right thing.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117632)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2550 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU73Qm0YYNi52MOFI-:-NoUFUyV443qeimJPnTG:b-com6rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb0e6c36d20cc50636e36f3211a0f6ac6367e26/pkg/kv/kvclient/kvcoord/dist_sender.go#L2550)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> "If not, we loop around and try the next replica"

Consider the following case: 

Range 10 has replicas on nodes 1, 2, 3, 4, 5; node 1 is the leaseholder but hasn't applied its lease yet. So, it continues to return the older leaseholder (say node 2) in its NLHE.

Say the client has updated leaseholder information. It contacts node 1 and gets stale lease information. It then ends up trying nodes 2, 3, 4, and 5 -- all of whom tell it that node 1 is the leaseholder. Finally, it exits this loop, evicts its range descriptor, and repeats the whole process again (until node 1 applies its lease). 

I think we're running into the issue https://github.com/cockroachdb/cockroach/pull/32877 was trying to solve again. Isn't the client better served simply backing off here, instead of trying nodes 2, 3, 4, and 5?
</blockquote></details>

EDIT: I don't think we're running into quite the same problem as #32877, in that we're not needlessly bouncing between the old and new leaseholder. My phrasing was a bit off there. However, we'll end up bouncing between all replicas, which is also bad.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
Agreed - this PR does make a substantial change in behavior. I feel fairly confident with a lot of the unit testing coverage, but I should probably run all the nightly roachtests on it as well prior to merging. Let me address all the other comments and will do that prior to a merge.

Note that I am going to rework some of these pieces early next week.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-:-NoXG4iSEGDgSHeXdY-s:bk222i7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 23 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU3U5TChLw94hMc0U5):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't think we should get rid of this metric. We should find room for it in the new structure. We've previously found this metric to be useful in support escalations where tail latency is high because a leaseholder is slow in applying its lease.
</blockquote></details>

Agreed - let me change the metric to be around the looping in `sendPartialBatch`. The nice thing with that change is it will "actually be accurate" vs today it is only counting the inner loop.
___
*[`-- commits` line 38 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU4OC-D3QnTEwA1IDV):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

See my previous comments -- we should still find a place for this metric.
</blockquote></details>

ack
___
*[`-- commits` line 51 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoUBc4a5OJwQ44K_6E8):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We're using a sync.Pool for the replicas slice on the transport already. Do we still expect to see major benefits?
</blockquote></details>

I'm planning on removing that in a future commit. Now that this is no longer mutated, it can be passed by reference and shared between transports. That removes the need for the `sync.Pool` completely. That said, I don't have a great perf test for this yet, but with the other refactoring it should be much easier to build one.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 251 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU4WhnE2515wNKst9o:-NoXH1DI1nxWGeOcoies:bzgqugg) ([raw file](https://github.com/cockroachdb/cockroach/blob/67664ad7a73b36145e4c48c14794636ef225bb9a/pkg/kv/kvclient/kvcoord/dist_sender.go#L251)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Getting rid of this seems like a downside of this approach. The rationale for this limit was explained in this comment: 

```
// If the leaseholder is the replica that we've just tried, and  
// we've tried this replica a bunch of times already, let's move on  
// and not try it again. This prevents us getting stuck on a replica  
// that we think has the lease but keeps returning redirects to us  
// (possibly because it hasn't applied its lease yet). Perhaps that  
// lease expires and someone else gets a new one, so by moving on we  
// get out of possibly infinite loops.
```

Are we no longer worried about this?
</blockquote></details>

Correct, this is no longer a concern. Previously it would "get stuck" retrying the leaseholder. This is bad for two reasons. First if the lease moved, but the "old leaseholder" didn't know it would just loop around for a long time. Second and more importantly, with the partial partition work we want to have distSender move onto different stores on any failure. We don't get stuck here because we always try other replicas. There is a small cost to doing that, but this is a rare situation so it shouldn't affect performance in normal cases.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 1892 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU5JBD6Ld2zfAzRCL3:-NoXISjTAKQvcE98EvDH:b6wa4rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb0e6c36d20cc50636e36f3211a0f6ac6367e26/pkg/kv/kvclient/kvcoord/dist_sender.go#L1892)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Isn't pErr nil here? Should this be err?
</blockquote></details>

Good point! I incorrectly changed this. I changed to be the correct wrapping of the underlying error.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2550 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU73Qm0YYNi52MOFI-:-NoXIbM86Mi8u4sxVtfu:b-ji36qc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb0e6c36d20cc50636e36f3211a0f6ac6367e26/pkg/kv/kvclient/kvcoord/dist_sender.go#L2550)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

EDIT: I don't think we're running into quite the same problem as #32877, in that we're not needlessly bouncing between the old and new leaseholder. My phrasing was a bit off there. However, we'll end up bouncing between all replicas, which is also bad.
</blockquote></details>

There is still the backoff, it is just in the `sendPartialBatch` rather than in `sendToReplicas`. As mentioned above, if we are hitting this case (leaseholder doesn't know they are the leaseholder), it is better to try other replicas as well. The previous behavior was pretty bad as well, since there were "nested retry" loops, it would hit the leaseholder much more than it does now. Say the backoffs are 1, 2, 4, ... 1024ms each and it took 10s for the leaseholder to get the lease. Then with the previous behavior it would have started the outer loop and try the inner loop 10 times (in ~2s), then it would restart the outer loop with a 1ms delay and retry the inner loop 10 times again. Then a 2ms outer loop delay and 10 more times on the inner loop. Now we correctly back off all the way to 10s, but more importantly try other replicas as well in case the lease moved to one of them. The reason this worked before was a little "lucky" in that when we went to the outer loop we would "forget" who the leaseholder was and try a random replica first, so if the lease did move then it could figure that out, but the timing and logic was "lucky" that it worked, and almost impossible to reason about.  
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1038 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoTj6JI85Vm5GuUN9vz:-NoXJlhzEhzXFa8kZSwb:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2c487e9e2533a0eea6b72a22f47e6367a4ebc3/pkg/kv/kvclient/rangecache/range_cache.go#L1038)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: s/its/it's/g
</blockquote></details>

Fixed
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1056 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU-snX0Lo7immOASES:-NoXKP3j7Z53-X_eqI_I:bf3e5du) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2c487e9e2533a0eea6b72a22f47e6367a4ebc3/pkg/kv/kvclient/rangecache/range_cache.go#L1056)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

> "the descriptors are the same"

Naviely, reading the code, one would assume the descriptor can be older as well. However that's not possible because the cache doesn't go backwards, and the supplied descriptor also came from the cache. So it must be equal to the one in the cache. This is explained in the comment below, but we should move that comment above this one.
</blockquote></details>

moved up and cleaned these comments up a little bit.
___
*[`pkg/kv/kvclient/rangecache/range_cache.go` line 1058 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoTtV6SEZHdm8Z5CDpx:-NoXKQ3v3BgpQe3I2OBM:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e2c487e9e2533a0eea6b72a22f47e6367a4ebc3/pkg/kv/kvclient/rangecache/range_cache.go#L1058)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: if/If/g and wrap comments to 80 chars.
</blockquote></details>

fixed
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go` line 259 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU9Im48-k33LwzHe8G:-NoXKiR68d9CFiPcJzmB:b-iqfpqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/67664ad7a73b36145e4c48c14794636ef225bb9a/pkg/kv/kvclient/kvcoord/dist_sender_ambiguous_test.go#L259)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is this happening because we're swallowing an error and not correctly wrapping it somewhere?
</blockquote></details>

Correct, it doesn't wrap exactly the same way anymore. Specifically the change in 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 841 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU3rY5CX44FbyAVqxf:-NoXLUHpD1_chyMC8FUC:b-yv5xzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/67664ad7a73b36145e4c48c14794636ef225bb9a/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L841)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is this test meaningful if we're getting rid of this?
</blockquote></details>

Let me add the metric back in and it will be useful again.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 919 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/117632#-NoU4BjYCH5egF-2Bll6:-NoXKxnE4Agkuu7ZZ6o0:b-fvfota) ([raw file](https://github.com/cockroachdb/cockroach/blob/67664ad7a73b36145e4c48c14794636ef225bb9a/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L919)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Once we find a place for this metric in the new structure, keeping this test around with the assertion will be useful in convincing us we're doing the right thing.
</blockquote></details>

Makes sense - I will recover this.


<!-- Sent from Reviewable.io -->
